### PR TITLE
login-logout

### PR DESCRIPTION
**Pull request by TrucLuu**

### CODE REVIEW

File: app/assets/javascripts/users.coffee [hunk 1]
```diff
@@ -0,0 +1,3 @@
+# Place all the behaviors and hooks related to the matching controller here.
+# All this logic will automatically be available in application.js.
+# You can use CoffeeScript in this file: http://coffeescript.org/
```
Comment (phamkykhoi) [inline app/assets/javascripts/users.coffee]: Remove this file

File: app/assets/javascripts/session.coffee [hunk 2]
```diff
@@ -0,0 +1,3 @@
+# Place all the behaviors and hooks related to the matching controller here.
+# All this logic will automatically be available in application.js.
+# You can use CoffeeScript in this file: http://coffeescript.org/
```
Comment (phamkykhoi) [inline app/assets/javascripts/session.coffee]: Remove this file

File: app/assets/stylesheets/users.scss [hunk 3]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the users controller here.
+// They will automatically be included in application.css.
+// You can use Sass (SCSS) here: http://sass-lang.com/
```
Comment (phamkykhoi) [inline app/assets/stylesheets/users.scss]: Remove this file

File: app/helpers/users_helper.rb [hunk 4]
```diff
@@ -0,0 +1,8 @@
+module UsersHelper
+  def gravatar_for user, size: 100
+  	return '' unless user
```
Comment (phamkykhoi) [inline app/helpers/users_helper.rb]: Indent

File: app/helpers/sessions_helper.rb [hunk 5]
```diff
@@ -0,0 +1,37 @@
+module SessionsHelper
+	def log_in user
```
Comment (phamkykhoi) [inline app/helpers/sessions_helper.rb]: Indent

File: app/assets/stylesheets/session.scss [hunk 6]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the Session controller here.
+// They will automatically be included in application.css.
+// You can use Sass (SCSS) here: http://sass-lang.com/
```
Comment (phamkykhoi) [inline app/assets/stylesheets/session.scss]: Remove this file

File: app/controllers/sessions_controller.rb [hunk 7]
```diff
@@ -0,0 +1,21 @@
+class SessionsController < ApplicationController
+  def new
+
```
Comment (phamkykhoi) [inline app/controllers/sessions_controller.rb]: Remove line empty in action

File: app/controllers/sessions_controller.rb [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+  end
+
+  def create
+  	user = User.find_by(email: params[:session][:email].downcase)
```
Comment (phamkykhoi) [inline app/controllers/sessions_controller.rb]: Indent

Comment (phamkykhoi) [inline app/controllers/sessions_controller.rb]: Remove ( )

File: app/controllers/sessions_controller.rb [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+  	if user && user.authenticate(params[:session][:password])
```
Comment (phamkykhoi) [inline app/controllers/sessions_controller.rb]: Indent

File: app/controllers/sessions_controller.rb [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+  		log_in user
```
Comment (phamkykhoi) [inline app/controllers/sessions_controller.rb]: Indent

File: app/controllers/sessions_controller.rb (hunk 9 above)
Comment (phamkykhoi) [inline app/controllers/sessions_controller.rb]: Remove ( )

File: app/controllers/sessions_controller.rb [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+  		redirect_back_or user
```
Comment (phamkykhoi) [inline app/controllers/sessions_controller.rb]: Indent

File: app/controllers/sessions_controller.rb [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+  	else
```
Comment (phamkykhoi) [inline app/controllers/sessions_controller.rb]: Indent

File: app/controllers/sessions_controller.rb [hunk 13]
```diff
@@ -0,0 +1,21 @@
...
+  	  flash[:error] = 'Invalid Email/Password combination'
```
Comment (phamkykhoi) [inline app/controllers/sessions_controller.rb]: Indent

Comment (phamkykhoi) [inline app/controllers/sessions_controller.rb]: ' ' =>  ""
use I18n

File: app/controllers/sessions_controller.rb [hunk 14]
```diff
@@ -0,0 +1,21 @@
...
+  	  render 'new'
```
Comment (phamkykhoi) [inline app/controllers/sessions_controller.rb]: ' ' =>  " "

File: app/controllers/sessions_controller.rb [hunk 15]
```diff
@@ -0,0 +1,21 @@
...
+  		log_in user
+  		redirect_back_or user
+  	else
+  	  flash[:error] = 'Invalid Email/Password combination'
+  	  render 'new'	
+  	end
+  end
+
+  def destroy
+  	log_out
```
Comment (phamkykhoi) [inline app/controllers/sessions_controller.rb]: Indent

File: app/controllers/users_controller.rb [hunk 16]
```diff
@@ -0,0 +1,43 @@
+class UsersController < ApplicationController
+  before_action :logged_in_user, only: [:edit, :update]
+  before_action :correct_user,   only: [:edit, :update]
```
Comment (phamkykhoi) [inline app/controllers/users_controller.rb]: Remove space ( alway 2 space )

File: app/controllers/users_controller.rb [hunk 17]
```diff
@@ -0,0 +1,43 @@
...
+  end
+
+  def edit
+  end
+
+  def create
+    @user = User.new user_params
+    if @user.save
+      log_in @user
+      flash[:success] = "Welcome to the E-Learning!"
```
Comment (phamkykhoi) [inline app/controllers/users_controller.rb]: I18n

File: app/controllers/users_controller.rb [hunk 18]
```diff
@@ -0,0 +1,43 @@
...
+      redirect_to @user
+    else
+      render 'new'
```
Comment (phamkykhoi) [inline app/controllers/users_controller.rb]: ' ' => " "

File: app/controllers/users_controller.rb [hunk 19]
```diff
@@ -0,0 +1,43 @@
...
+    end
+  end
+
+  def update
+  end
+
+  private
+  def user_params
+    param.require(:user).permit(:name, :email, :password,
```
Comment (phamkykhoi) [inline app/controllers/users_controller.rb]: Remove ( )
Indent

File: app/helpers/sessions_helper.rb (hunk 5 above)
Comment (phamkykhoi) [inline app/helpers/sessions_helper.rb]: Indent

File: app/helpers/sessions_helper.rb [hunk 20]
```diff
@@ -0,0 +1,37 @@
...
+		session[:user_id] = user.id
```
Comment (phamkykhoi) [inline app/helpers/sessions_helper.rb]: Indent

File: app/helpers/sessions_helper.rb [hunk 21]
```diff
@@ -0,0 +1,37 @@
...
+    session.delete(:forwarding_url)
+  end
+  def store_location
+    session[:forwarding_url] = request.url if request.get?
+  end
+
+  def logged_in_user
+    unless logged_in?
+      store_location
+      flash[:danger] = "Please log in."
```
Comment (phamkykhoi) [inline app/helpers/sessions_helper.rb]: I18n

File: app/helpers/users_helper.rb [hunk 22]
```diff
@@ -0,0 +1,8 @@
...
+	  gravatar_id = Digest::MD5::hexdigest user.email.downcase
+	  gravatar_url = "https://secure.gravatar.com/avatar/#{gravatar_id}.png?s=#{size}"
```
Comment (phamkykhoi) [inline app/helpers/users_helper.rb]: Indent

File: app/helpers/users_helper.rb [hunk 23]
```diff
@@ -0,0 +1,8 @@
...
+	  image_tag gravatar_url, alt: user.name, class: "gravatar"
```
Comment (phamkykhoi) [inline app/helpers/users_helper.rb]: Indent

File: app/helpers/users_helper.rb (hunk 4 above)
Comment (phamkykhoi) [inline app/helpers/users_helper.rb]: ' ' => " "

File: app/views/layouts/_header.html.erb [hunk 24]
```diff
@@ -5,6 +5,24 @@
...
+                <%= link_to "Log out", logout_path, method: "delete" %>
+              </li>
+            </ul>
+          </li>
+        <% else %>
+          <li><%= link_to t("login"), login_path %></li>
+        <% end %>
       </ul>
     </nav>
   </div>
```
Comment (phamkykhoi) [inline app/views/layouts/_header.html.erb]: Check Indent this file

File: app/views/sessions/new.html.erb [hunk 25]
```diff
@@ -0,0 +1,19 @@
+<% provide(:title, "Log in") %>
```
Comment (phamkykhoi) [inline app/views/sessions/new.html.erb]: I18n

File: app/views/sessions/new.html.erb [hunk 26]
```diff
@@ -0,0 +1,19 @@
...
+
+      <%= f.label :password %>
+      <%= f.password_field :password, class: 'form-control' %>
+
+      <%= f.submit "Log in", class: "btn btn-primary" %>
+    <% end %>
+
+    <p>New user? <%= link_to "Sign up now!", "#" %></p>
+  </div>
+</div>
```
Comment (phamkykhoi) [inline app/views/sessions/new.html.erb]: Check indent this file

File: app/views/users/index.html.erb [hunk 27]
```diff
@@ -0,0 +1,12 @@
+<% provide(:title, 'All users') %>
```
Comment (phamkykhoi) [inline app/views/users/index.html.erb]: I18n 
Check indent this  file

File: config/routes.rb [hunk 28]
```diff
@@ -3,4 +3,11 @@
   get "help" => "static_pages#help"
   get "about" => "static_pages#about"
   get "contact" => "static_pages#contact"
+
+  resources :users
+  resources :sessions, only: [:new, :create, :destroy]
+
+  get "login" => 'sessions#new'
```
Comment (phamkykhoi) [inline config/routes.rb]: ' ' => " "

File: config/routes.rb [hunk 29]
```diff
@@ -3,4 +3,11 @@
...
+  post "login" => 'sessions#create'
```
Comment (phamkykhoi) [inline config/routes.rb]: ' ' => " "

File: config/routes.rb [hunk 30]
```diff
@@ -3,4 +3,11 @@
...
+  delete "logout" => 'sessions#destroy'
```
Comment (phamkykhoi) [inline config/routes.rb]: ' ' => " "

File: app/views/users/index.html.erb [hunk 31]
```diff
@@ -0,0 +1,12 @@
...
+<h1>
+  All users
+</h1>
+<ul class="users">
+  <% @users.each do |user| %>
+    <li>
+      <%= gravatar_for user, size: 50 %>
```
Comment (phamkykhoi) [inline app/views/users/index.html.erb]: size: 50
use Setting gem

File: app/controllers/users_controller.rb [hunk 32]
```diff
@@ -0,0 +1,43 @@
...
+    param.require(:user).permit(:name, :email, :password, 
+                                :password_confirmation)
+  end
+
+  def set_user
+    @user = User.find(params[:id])
+  end
+
+  def correct_user
+      @users = User.find(params[:id])
```
Comment (phamkykhoi) [inline app/controllers/users_controller.rb]: Indent

File: app/controllers/users_controller.rb [hunk 33]
```diff
@@ -0,0 +1,43 @@
...
+                                :password_confirmation)
+  end
+
+  def set_user
+    @user = User.find(params[:id])
+  end
+
+  def correct_user
+      @users = User.find(params[:id])  
+      redirect_to(root_url) unless @user == current_user?(@user)
```
Comment (phamkykhoi) [inline app/controllers/users_controller.rb]: Indent

Comment (phamkykhoi) [inline app/controllers/users_controller.rb]: current_user?(@user)  =>  Remove ( )

File: app/models/user.rb [hunk 34]
```diff
@@ -7,5 +7,18 @@ class User < ActiveRecord::Base
...
-  has_many :followers, through: :passive_relationships, source: :follower
+  has_many :followers, through: :passive_relationships, source: :follower,
+                       dependent: :destroy
+ 	validates :name, presence: true, length: { maximum: 50 }
+ 	VALID_EMAIL_REGEX = /\A[\w+\-.]+@[a-z\d\-.]+\.[a-z]+\z/i
+ 	validates :email, presence: true, length: { maximum: 255 },
+                                    format: { with: VALID_EMAIL_REGEX },
+                                    uniqueness: { case_sensitive: false }
+  has_secure_password
+  validates :password, length: { minimum: 6 }
```
Comment (phamkykhoi) [inline app/models/user.rb]: Remove space  { }

File: app/models/user.rb [hunk 35]
```diff
@@ -7,5 +7,18 @@ class User < ActiveRecord::Base
...
                                   dependent: :destroy
   has_many :following, through: :active_relationships, source: :followed
-  has_many :followers, through: :passive_relationships, source: :follower
+  has_many :followers, through: :passive_relationships, source: :follower,
+                       dependent: :destroy
+ 	validates :name, presence: true, length: { maximum: 50 }
+ 	VALID_EMAIL_REGEX = /\A[\w+\-.]+@[a-z\d\-.]+\.[a-z]+\z/i
+ 	validates :email, presence: true, length: { maximum: 255 },
+                                    format: { with: VALID_EMAIL_REGEX },
+                                    uniqueness: { case_sensitive: false }
```
Comment (phamkykhoi) [inline app/models/user.rb]: Remove space { }

File: app/models/user.rb [hunk 36]
```diff
@@ -7,5 +7,18 @@ class User < ActiveRecord::Base
                                   foreign_key: "followed_id",
                                   dependent: :destroy
   has_many :following, through: :active_relationships, source: :followed
-  has_many :followers, through: :passive_relationships, source: :follower
+  has_many :followers, through: :passive_relationships, source: :follower,
+                       dependent: :destroy
+ 	validates :name, presence: true, length: { maximum: 50 }
+ 	VALID_EMAIL_REGEX = /\A[\w+\-.]+@[a-z\d\-.]+\.[a-z]+\z/i
+ 	validates :email, presence: true, length: { maximum: 255 },
+                                    format: { with: VALID_EMAIL_REGEX },
```
Comment (phamkykhoi) [inline app/models/user.rb]: Remove space { }

File: app/models/user.rb [hunk 37]
```diff
@@ -7,5 +7,18 @@ class User < ActiveRecord::Base
                                   foreign_key: "followed_id",
                                   dependent: :destroy
   has_many :following, through: :active_relationships, source: :followed
-  has_many :followers, through: :passive_relationships, source: :follower
+  has_many :followers, through: :passive_relationships, source: :follower,
+                       dependent: :destroy
+ 	validates :name, presence: true, length: { maximum: 50 }
+ 	VALID_EMAIL_REGEX = /\A[\w+\-.]+@[a-z\d\-.]+\.[a-z]+\z/i
+ 	validates :email, presence: true, length: { maximum: 255 },
```
Comment (phamkykhoi) [inline app/models/user.rb]: Remove space { }

File: app/models/user.rb (hunk 34 above)
Comment (phamkykhoi) [inline app/models/user.rb]: Remove space { }

File: app/models/user.rb [hunk 38]
```diff
@@ -7,5 +7,18 @@ class User < ActiveRecord::Base
...
+                                    format: { with: VALID_EMAIL_REGEX },
+                                    uniqueness: { case_sensitive: false }
+  has_secure_password
+  validates :password, length: { minimum: 6 }
+  def User.digest string
```
Comment (phamkykhoi) [inline app/models/user.rb]: above is line empty

File: app/helpers/sessions_helper.rb [hunk 39]
```diff
@@ -0,0 +1,37 @@
...
+  end
+
+  def logged_in?
+    !current_user.nil?
+  end
+
+  def current_user? user
+    user == current_user
+  end
+  def redirect_back_or default
```
Comment (phamkykhoi) [inline app/helpers/sessions_helper.rb]: Insert line empty above

File: app/views/layouts/_header.html.erb [hunk 40]
```diff
@@ -5,6 +5,24 @@
         <li><%= link_to t("home"), root_path %></li>
         <li><%= link_to t("help"), help_path %></li>
         <li><%= link_to t("contact"), contact_path %></li>
+        <% if logged_in? %>
+          <li><%= link_to "Users", users_path %></li>
```
Comment (phamkykhoi) [inline app/views/layouts/_header.html.erb]: I18n

File: app/views/layouts/_header.html.erb [hunk 41]
```diff
@@ -5,6 +5,24 @@
...
+          <li class="dropdown">
+            <a href="#" class="dropdown-toggle" data-toggle="dropdown">
+              Account <b class="caret"></b>
+            </a>
+            <ul class="dropdown-menu">
+              <li><%= link_to "Profile", current_user %></li>
+              <li><%= link_to "Settings", edit_user_path(current_user) %></li>
+              <li class="divider"></li>
+              <li>
+                <%= link_to "Log out", logout_path, method: "delete" %>
```
Comment (phamkykhoi) [inline app/views/layouts/_header.html.erb]: I18n

File: app/views/sessions/new.html.erb [hunk 42]
```diff
@@ -0,0 +1,18 @@
+<% provide(:title, t(:login)[:button]) %>
+<h1>Log in</h1>
```
Comment (phamkykhoi) [inline app/views/sessions/new.html.erb]: I18n

File: app/views/users/index.html.erb [hunk 43]
```diff
@@ -0,0 +1,12 @@
+<% provide(:title, t(:users)) %>
+<h1>
+  All users
```
Comment (phamkykhoi) [inline app/views/users/index.html.erb]: I18n

File: config/routes.rb [hunk 44]
```diff
@@ -3,4 +3,11 @@
...
   get "about" => "static_pages#about"
   get "contact" => "static_pages#contact"
+
+  resources :users
+  resources :sessions, only: [:new, :create, :destroy]
+
+  get "login" => "sessions#new"
+  post "login" => "sessions#create"
+  delete "logout" => "sessions#destroy"
 end
```
Comment (phamkykhoi) [inline config/routes.rb]: get "login" => "sessions#new"
post "login" => "sessions#create"
delete "logout" => "sessions#destroy"
dupplicate with  resources :sessions, only: [:new, :create, :destroy]

File: app/controllers/users_controller.rb [hunk 45]
```diff
@@ -0,0 +1,47 @@
...
+  end
+
+  def create
+    @user = User.new user_params
+    if @user.save
+      log_in @user
+      flash[:success] = t(:login)[:success]
+      redirect_to @user
+    else
+      render :new
```
Comment (phamkykhoi) [inline app/controllers/users_controller.rb]: render :new ==> render "new"

File: app/controllers/users_controller.rb [hunk 46]
```diff
@@ -0,0 +1,47 @@
...
+    end
+  end
+
+  def update
+  end
+
+  private
+  def user_params
+    param.require :user.permit(
```
Comment (phamkykhoi) [inline app/controllers/users_controller.rb]: remove ( ) after permit

Comment (phamkykhoi) [inline app/controllers/users_controller.rb]: :name,
:email,
:password,
:password_confirmation
=> :name, :email, :password, :password_confirmation

File: app/models/user.rb [hunk 47]
```diff
@@ -7,5 +9,24 @@ class User < ActiveRecord::Base
                                   foreign_key: "followed_id",
                                   dependent: :destroy
   has_many :following, through: :active_relationships, source: :followed
-  has_many :followers, through: :passive_relationships, source: :follower
+  has_many :followers, through: :passive_relationships, source: :follower,
+                       dependent: :destroy
+  has_secure_password
+
+  validates :name, presence: true, length: {maximum: 50}
```
Comment (phamkykhoi) [inline app/models/user.rb]: Move Settings

File: app/models/user.rb [hunk 48]
```diff
@@ -7,5 +9,24 @@ class User < ActiveRecord::Base
...
+  validates :email, presence: true, length: {maximum: 255},
```
Comment (phamkykhoi) [inline app/models/user.rb]: Move Settings

File: app/models/user.rb [hunk 49]
```diff
@@ -7,5 +9,24 @@ class User < ActiveRecord::Base
...
+                    format: {with: VALID_EMAIL_REGEX},
+                    uniqueness: {case_sensitive: false},
+                    if: "new_record?"
+  validates :password, length: {minimum: 6}, if: "password_set?"
```
Comment (phamkykhoi) [inline app/models/user.rb]: Move Settings

File: app/controllers/sessions_controller.rb [hunk 50]
```diff
@@ -0,0 +1,20 @@
...
+  end
+
+  def create
+    user = User.find_by email: params[:session][:email].downcase
+    if user && user.authenticate(params[:session][:password])
+      log_in user
+      redirect_back_or user
+    else
+      flash[:error] = t(:login)[:failed]
+      render "new "
```
Comment (phamkykhoi) [inline app/controllers/sessions_controller.rb]: Remove space

File: app/helpers/sessions_helper.rb [hunk 51]
```diff
@@ -0,0 +1,39 @@
+module SessionsHelper
+  def log_in user
+    session[:user_id] = user.id
+  end
+
+  def log_out
+    session.delete(:user_id)
```
Comment (phamkykhoi) [inline app/helpers/sessions_helper.rb]: Thử bỏ ( ) after delete xem có được ko

File: app/helpers/sessions_helper.rb [hunk 52]
```diff
@@ -0,0 +1,39 @@
...
+    !current_user.nil?
+  end
+
+  def current_user? user
+    user == current_user
+  end
+
+  def redirect_back_or default
+    redirect_to(session[:forwarding_url] || default)
+    session.delete(:forwarding_url)
```
Comment (phamkykhoi) [inline app/helpers/sessions_helper.rb]: Thử bỏ ( ) after delete xem có được ko

File: app/views/sessions/new.html.erb [hunk 53]
```diff
@@ -0,0 +1,18 @@
+<% provide(:title, t(:login)[:button]) %>
+<h1><%= t(:login)[:button] %></h1>
+
+<div class="row">
+  <div class="col-md-6 col-md-offset-3">
+    <%= form_for(:session, url: login_path) do |f| %>
+      <%= f.label :email %>
+      <%= f.email_field :email, class: 'form-control' %>
```
Comment (phamkykhoi) [inline app/views/sessions/new.html.erb]: ' ' => " "

File: app/views/sessions/new.html.erb [hunk 54]
```diff
@@ -0,0 +1,18 @@
...
+
+      <%= f.label :password %>
+      <%= f.password_field :password, class: 'form-control' %>
```
Comment (phamkykhoi) [inline app/views/sessions/new.html.erb]: ' ' => " "

File: app/views/sessions/new.html.erb [hunk 55]
```diff
@@ -0,0 +1,18 @@
+<% provide(:title, t(:login)[:button]) %>
+<h1><%= t(:login)[:button] %></h1>
+
+<div class="row">
+  <div class="col-md-6 col-md-offset-3">
+    <%= form_for(:session, url: login_path) do |f| %>
```
Comment (phamkykhoi) [inline app/views/sessions/new.html.erb]: Remove ( ) after form_for

File: app/views/layouts/_header.html.erb [hunk 56]
```diff
@@ -5,6 +5,24 @@
         <li><%= link_to t("home"), root_path %></li>
         <li><%= link_to t("help"), help_path %></li>
         <li><%= link_to t("contact"), contact_path %></li>
+        <% if logged_in? %>
+          <li><%= link_to t(:login)[:user], users_path %></li>
+          <li class="dropdown">
+            <a href="#" class="dropdown-toggle" data-toggle="dropdown">
+              Account <b class="caret"></b>
```
Comment (phamkykhoi) [inline app/views/layouts/_header.html.erb]: I18n

File: app/models/user.rb [hunk 57]
```diff
@@ -1,4 +1,6 @@
 class User < ActiveRecord::Base
+  VALID_EMAIL_REGEX = /\A[\w+\-.]+@[a-z\d\-.]+\.[a-z]+\z/i
```
Comment (phamkykhoi) [inline app/models/user.rb]: Remove Settings

File: app/controllers/users_controller.rb [hunk 58]
```diff
@@ -0,0 +1,42 @@
...
+    param.require :user.permit :name,:email,:password,:password_confirmation
+  end
+
+  def set_user
+    @user = User.find params[:id]
+  end
+
+  def correct_user
+    @users = User.find params[:id]
+    redirect_to(root_url) unless current_user?(@user)
```
Comment (phamkykhoi) [inline app/controllers/users_controller.rb]: Remove  () of (@user)

File: app/controllers/users_controller.rb [hunk 59]
```diff
@@ -0,0 +1,42 @@
+class UsersController < ApplicationController
+  before_action :logged_in_user, only: [:edit, :update]
+  before_action :correct_user, only: [:edit, :update]
+  before_action :set_user, only: [:show, :edit, :update]
+  def index
```
Comment (phamkykhoi) [inline app/controllers/users_controller.rb]: Should insert empty line above

File: app/views/layouts/_header.html.erb [hunk 60]
```diff
@@ -5,6 +5,24 @@
...
+        <% if logged_in? %>
+          <li><%= link_to t(:login)[:user], users_path %></li>
+          <li class="dropdown">
+            <a href="#" class="dropdown-toggle" data-toggle="dropdown">
+              <%= t(:login)[:account] %> <b class="caret"></b>
+            </a>
+            <ul class="dropdown-menu">
+              <li><%= link_to t(:login)[:profile], current_user %></li>
+              <li><%= link_to t(:login)[:setting], edit_user_path(current_user) %></li>
+              <li class="divider"></li>
```
Comment (phamkykhoi) [inline app/views/layouts/_header.html.erb]: What is this tag li ?

File: app/helpers/sessions_helper.rb [hunk 61]
```diff
@@ -0,0 +1,39 @@
...
+  def logged_in?
+    !current_user.nil?
+  end
+
+  def current_user? user
+    user == current_user
+  end
+
+  def redirect_back_or default
+    redirect_to(session[:forwarding_url] || default)
```
Comment (dieunb) [inline app/helpers/sessions_helper.rb]: remove `()`

File: app/views/sessions/new.html.erb [hunk 62]
```diff
@@ -0,0 +1,18 @@
+<% provide(:title, t(:login)[:button]) %>
```
Comment (dieunb) [inline app/views/sessions/new.html.erb]: remove `()`

File: app/views/users/index.html.erb [hunk 63]
```diff
@@ -0,0 +1,12 @@
+<% provide(:title, t("users")) %>
```
Comment (dieunb) [inline app/views/users/index.html.erb]: remove `()`

File: app/views/users/show.html.erb [hunk 64]
```diff
@@ -0,0 +1,11 @@
+<% provide(:title, @user.name) %>
```
Comment (dieunb) [inline app/views/users/show.html.erb]: remove `()`

File: app/controllers/users_controller.rb [hunk 65]
```diff
@@ -1,20 +1,51 @@
...
     end
   end
 
+  def update
+  end
+
   private
   def user_params
-    params.require(:user).permit :name, :email, :password, :password_confirmation
+    params.require(:user).permit(
```
Comment (phamkykhoi) [inline app/controllers/users_controller.rb]: Remove ()

File: app/controllers/users_controller.rb [hunk 66]
```diff
@@ -1,20 +1,51 @@
...
+      :name,
+      :email,
+      :password,
+      :password_confirmation
```
Comment (phamkykhoi) [inline app/controllers/users_controller.rb]: + :name,
+ :email,
+ :password,
+ :password_confirmation
Move once line

File: Gemfile [hunk 67]
```diff
@@ -10,7 +10,8 @@ gem "jquery-rails"
 gem "turbolinks"
 gem "jbuilder", "~> 2.0"
 gem "sdoc", "~> 0.4.0", group: :doc
-gem "bcrypt", "~> 3.1.7"
+gem "bcrypt", '~> 3.1.7'
```
Comment (dieunb) [inline Gemfile]: use `"` instead `'`

File: app/helpers/sessions_helper.rb [hunk 68]
```diff
@@ -0,0 +1,39 @@
...
+  end
+  
+  def store_location
+    session[:forwarding_url] = request.url if request.get?
+  end
+
+  def logged_in_user
+    unless logged_in?
+      store_location
+      flash[:danger] = t(:login)[:require]
```
Comment (dieunb) [inline app/helpers/sessions_helper.rb]: `t "login.require"`

File: app/views/layouts/_header.html.erb [hunk 69]
```diff
@@ -5,6 +5,23 @@
         <li><%= link_to t("home"), root_path %></li>
         <li><%= link_to t("help"), help_path %></li>
         <li><%= link_to t("contact"), contact_path %></li>
+        <% if logged_in? %>
+          <li><%= link_to t(:login)[:user], users_path %></li>
```
Comment (dieunb) [inline app/views/layouts/_header.html.erb]: `t("login.user")`

File: app/views/layouts/_header.html.erb [hunk 70]
```diff
@@ -5,6 +5,23 @@
...
+          <li class="dropdown">
+            <a href="#" class="dropdown-toggle" data-toggle="dropdown">
+              <%= t(:login)[:account] %> <b class="caret"></b>
```
Comment (dieunb) [inline app/views/layouts/_header.html.erb]: `t("login.account"`

File: app/views/layouts/_header.html.erb [hunk 71]
```diff
@@ -5,6 +5,23 @@
...
+            </a>
+            <ul class="dropdown-menu">
+              <li><%= link_to t(:login)[:profile], current_user %></li>
```
Comment (dieunb) [inline app/views/layouts/_header.html.erb]: `t("login.profile")`

File: app/views/layouts/_header.html.erb [hunk 72]
```diff
@@ -5,6 +5,23 @@
...
+              <li><%= link_to t(:login)[:setting], edit_user_path(current_user) %></li>
```
Comment (dieunb) [inline app/views/layouts/_header.html.erb]: `t("login.setting")`

File: app/views/layouts/_header.html.erb [hunk 73]
```diff
@@ -5,6 +5,23 @@
...
+              <li>
+                <%= link_to t(:login)[:logout], logout_path, method: "delete" %>
```
Comment (dieunb) [inline app/views/layouts/_header.html.erb]: `t("login.logout")`

File: app/views/layouts/_header.html.erb [hunk 74]
```diff
@@ -5,6 +5,23 @@
...
+              </li>
+            </ul>
+          </li>
+        <% else %>
+          <li><%= link_to t(:login)[:button], login_path %></li>
```
Comment (dieunb) [inline app/views/layouts/_header.html.erb]: `t("login.button")`

**Pull request merged**